src/compiler.rkt

```diff
@@ -6,101 +6,106 @@
 
 (provide compile-specs compile-spec compile-progs compile-prog)
 
-;; Function calculates a precision for the operation
-;;    with respect to the given extra-precision (exponents from a previous run)
-(define (operator-precision extra-precision)
-  (min (*max-mpfr-prec*)
-       (+ extra-precision
-          (*ground-truth-extra-bits*)
-          (bf-precision))))
-
 (define (true-exponent x)
-  (+ (bigfloat-exponent x) (bigfloat-precision x)))
-
-(define (unbox-prec x)
-    (if (box? x)
-        (unbox x)
-        0))
+  (define exp (+ (bigfloat-exponent x) (bigfloat-precision x)))
+  (if (equal? exp -9223372036854775807)
+      (- (get-slack))  ; 0.bf
+      (if (or (< 1000000000 exp) (equal? exp -9223372036854775805))
+          (get-slack)  ; overflow/inf.bf
+          exp)))
 
 ;; Interpreter taking a narrow IR
 ;; ```
 ;; <prog> ::= #(<instr> ..+)
 ;; <instr> ::= '(<op-procedure> <index> ...)
-;; <op-procedure> ::= #(<operation> <working-precision> <extra-precision> <exponents-checkpoint>)
 ;; ```
 ;; where <index> refers to a previous virtual register.
 ;; Must also provide the input variables for the program(s)
 ;; as well as the indices of the roots to extract.
 ;; name ::= 'fl or 'ival
 (define (make-progs-interpreter name vars ivec roots)
-  (define vreg-count (+ (length vars) (vector-length ivec)))
+  (define varc (length vars))
+  (define vreg-count (+ varc (vector-length ivec)))
   (define vregs (make-vector vreg-count))
-
-  (define (tuning-filter instr)
-    (if (member
-         (vector-ref (car instr) 0)
-         (list ival-sin ival-cos ival-tan))
-        #t
-        #f))
+  (define vprecs (make-vector vreg-count))             ; vector that stores working precisions
+  (define vstart-precs (setup-vstart-precs ivec varc)) ; starting precisions for the tuning mode
```

Comment:
Hmm, but we here do it once, when inside backward pass it will have to be done every iteration. I hope that this function shouldn't be expensive, but I can move and try it in another place

Reply:
Yeah I'm not worried about expensive, it seems pretty cheap. It would just be nice if two similar functions could actually be one function, that is, less lines of code. It isn't that important.